core/src/client/async_client/mod.rs

```diff
@@ -726,95 +771,157 @@ async fn background_task<S, R>(
 }
 
-fn unparse_error(raw: &[u8]) -> Error {
-	let json = serde_json::from_slice::<serde_json::Value>(raw);
+struct ReadTaskParams<R: TransportReceiverT> {
+	receiver: R,
+	close_tx: mpsc::Sender<Result<(), Error>>,
+	to_send_task: mpsc::Sender<FrontToBack>,
+	manager: ThreadSafeRequestManager,
+	max_buffer_capacity_per_subscription: usize,
+}
 
-	let json_str = match json {
-		Ok(json) => serde_json::to_string(&json).expect("valid JSON; qed"),
-		Err(e) => e.to_string(),
+async fn read_task<R>(params: ReadTaskParams<R>)
+where
+	R: TransportReceiverT,
+{
+	let ReadTaskParams { receiver, close_tx, to_send_task, manager, max_buffer_capacity_per_subscription } = params;
+
+	let backend_event = futures_util::stream::unfold(receiver, |mut receiver| async {
+		let res = receiver.receive().await;
+		Some((res, receiver))
+	});
+
+	let closed = close_tx.closed();
+	let pending_unsubscribes = MaybePendingFutures::new();
+
+	tokio::pin!(backend_event, closed, pending_unsubscribes);
+
+	let rx_item = backend_event.next();
+	let mut rx_or_closed = future::select(rx_item, closed);
+
+	let res = loop {
+		match future::select(rx_or_closed, pending_unsubscribes.next()).await {
+			// New message received to process.
+			Either::Left((Either::Left((Some(msg), closed_fut)), _)) => {
+				match handle_backend_messages::<R>(Some(msg), &manager, max_buffer_capacity_per_subscription) {
+					Ok(Some(msg)) => {
+						pending_unsubscribes.push(to_send_task.send(msg));
+					}
+					Err(e) => {
+						tracing::error!("[backend]: Failed to read message: {e}");
+						break Err(e);
+					}
+					Ok(None) => (),
+				}
+				rx_or_closed = future::select(backend_event.next(), closed_fut);
+			}
+			// Unsubscribe completed.
+			Either::Right((Some(Ok(_)), rx_or_closed_fut)) => {
+				rx_or_closed = rx_or_closed_fut;
+			}
+			// Closed.
+			_ => break Ok(()),
+		};
 	};
 
-	Error::Custom(format!("Unparseable message: {json_str}"))
+	let _ = close_tx.send(res).await;
+}
+
+async fn wait_for_shutdown(
+	mut close_rx: mpsc::Receiver<Result<(), Error>>,
+	client_dropped: oneshot::Receiver<()>,
+	err_to_front: oneshot::Sender<Error>,
+) {
+	let rx_item = close_rx.recv();
+
+	tokio::pin!(rx_item);
+
+	// Send an error to the frontend if the send or receive task completed with an error.
+	if let Either::Left((Some(Err(err)), _)) = future::select(rx_item, client_dropped).await {
+		let _ = err_to_front.send(err);
+	}
+}
+
+/// A wrapper around `FuturesUnordered` that doesn't return `None` when it's empty.
+struct MaybePendingFutures<Fut> {
+	futs: FuturesUnordered<Fut>,
+}
+
+impl<Fut> MaybePendingFutures<Fut> {
+	fn new() -> Self {
+		Self { futs: FuturesUnordered::new() }
+	}
+
+	fn push(&mut self, fut: Fut) {
+		self.futs.push(fut);
+	}
+}
+
+impl<Fut: Future> Stream for MaybePendingFutures<Fut> {
+	type Item = Fut::Output;
+
+	fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+		if self.futs.is_empty() {
+			return Poll::Pending;
```

Comment:
At this point, the waker is not handed anywhere and so if any other futures are added, they will be ignored.

I _think_ that this is ok though because you just call `next()` on this in a loop, so it'll get polled at least once per loop anyway.

Reply:
I tried using pure `FutureUnordered` and then the select loop will be busy that it returns `Poll::Ready i.e None` when it's empty and steal CPU cycles from more important stuff, that's the reason why I added it...

Reply:
Yeah, if it's empty and futures are added those will read on the next iteration

Reply:
Ah I see, so when it's empty it'll spam `Poll::Ready(None)` whenever polled; yeah that sucks. Your version looks good to me offhand in the current code now, though it's a little scary that if used wrong it'll just `.await` forever, so a small footgun to be wary of for future us!

Reply:
I believe this should be fine for now, because we use the read_task to drive this:

```
read_task
	select (server_receiver, closed, pending_unsub)
		server_receiver -> if Ok(Some(m)) = handle_backend_messages
			pending_unsub.push( to_send_task.send(m) )
```

One case where this might be problematic is when:
- server_receiver and closed futures are not ready
 - the pending_unsub is empty and thus will return `Poll::Pending` and I believe will never woke up for the current iteration loop because there's nothing that will wake up the `cx.waker` again of the future
- then pending_unsub has some elements added to it

But since`pending_unsub` having elements added is dependent on `server_receiver` being triggered first; it should be fine here as long as we don't save the current `pending_unsub` future that returned `Pending` and instead call next directly in the loop `future::select(rx_or_closed, pending_unsubscribes.next())`

Let me know if my reasoning is right 🙏 

Indeed there might be dragons if we change something in the `read_task` (maybe a comment to not save `pending_unsub.next()` future is enough here).

We may be able to handle this by:
 - if futs.is_empty(), then we could save the cx
 - when adding a new future, if we have `Some(cx)` we could wake up the waker
 - preserve the same pending_unsub.next() between iterations similar to `let rx_item = backend_event.next();`
 
 Not quite sure if worth going that way

Reply:
yeah, you are correct.

Reply:
did what you suggested it didn't require much code but instead I had to clone the Waker because `Context` is !Send

Reply:
Ah that's what you did!